### ISSUE TITLE
Should ValueSet in p4info.proto have match_fields like Table does, instead of bitwidth?

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4runtime#135]
In p4runtime.proto, ValueSetEntry has a repeated match field of type FieldMatch, just like the TableEntry message does.

That seems reasonable to me.  What seems wrong is that in the p4info.proto, message Table has a repeated match_fields of type MatchField, but message ValueSet has bitwidth.

**Comment by antoninbas** [comment]
I think that at the time this was added, there was some debate going on about match types for value sets. Looking at the spec (https://p4.org/p4-spec/docs/P4-16-v1.1.0-spec.html#sec-value-set) I do not see any way to specify a match type. Which means that for me the client can specify an arbitrary match and it's up to the target to accept or reject it. It kind of mean that a decent implementation would hope TCAMs if possible in order to specify all match types (save for `RANGE` possibly). The P4Runtime spec should probably include some additional information though.

**Comment by jafingerhut** [comment]
Search in the language spec for the `@match` annotation, which I believe may be what you are looking for in terms of match_kind for value_set fields.

**Comment by antoninbas** [comment]
Thanks! It may be worth reaching out to a few people to see if we can break backward compatibility and replace bitwidth with `MatchField`. The alternative is to add a `MatchField.MatchType` field to `ValueSet`.

**Comment by jafingerhut** [comment]
Makes sense on reaching out.  However, with the current contents of P4Info and ValueSetEntry, I don't see how entries can be added to a value set at all.  If someone has figured out how to make the current thing work at all, then I guess breaking backwards compatibility would be an issue.  I do not understand how one can fill in the contents of a ValueSetEntry message _at all_ with the current message contents.  Where would you get the `field_id` values to put inside the `FieldMatch` messages with today's p4info contents?

**Comment by antoninbas** [comment]
The spec should say explicitly that the `field_id` is always unset (0) for a `ValueSetEntry` (right now that's implicit in the example). We are matching on a single element, whose type is determined by the type parameter of the P4 `value_set` objects.
In the spec example:
```
type: INSERT
entity {
  value_set_entry {
    value_set_id : 0x03000001
    match { exact { value: 0x22F3 } } }
    match { exact { value: 0x893B } } }
  }
}
```
We are configuring the value set to have 2 elements.

If the value set includes something more complicated than `bit<W>` or `int<W>`, then we run into the well-known limitation of `MatchField` being limited to simple bitstrings (instead of the general `P4Data`).
However, I wasn't expecting the P4 spec to enable specifying a different match type for each struct field when using a struct as the value set type parameter. I don't like how this turned out because the P4 spec doesn't force the programmer to use a struct, but at the same time, if one wants to specify match kind using `@match`, one *has* to use a struct (e.g. I don't think this can be done with a tuple). This means that each field in the struct would need to be a different `MatchField`...

So we have 2 options for P4Runtime 1.0:
 * only support P4 value sets parametrized by `bit<W>`, `int<W>` or a struct with a *single* field (which can be annotated with `@match`). We can mostly keep what we have today, but include the match kind in P4Info somehow.

* fix current support the proper way to match the P4 spec. The `ValueSet` and `ValueSetEntry` messages would probably need to look like this:
```
message ValueSet {
  Preamble preamble = 1;
  repeated MatchField match = 2;
  int32 size = 3;
}
```
```
message ValueSetEntry {
  message Member {
    repeated FieldMatch match = 1;
  }

  uint32 value_set_id = 1;
  repeated Member members = 2;
}
```

**Comment by jnfoster** [comment]
Can you summarize the changes you'd like to make to the P4 Spec?

**Comment by antoninbas** [comment]
@jnfoster I don't think there is any strong need to change anything in the spec at this stage.
But I have 2 questions / remarks:
 * I don't like the asymmetry between exact match and the other kinds of match. If I want to match on a bitstring, I can use `value_set<bit<W> >` if it is an exact match, but I have to wrap my `bit<W>` in a struct for other matches.
 * If the match semantics are as rich as a table, I'm not sure why we didn't go with a similar syntax.

But once again there is no need to change anything. In P4Runtime we are going with something very similar to table matches, which also works fine with the current P4 `value_set` specification.

**Comment by jafingerhut** [comment]
I agree that something like table definition syntax for a value_set would have been more "symmetric", given the similarity there.

Is there a desire that value_set's are able to be "looked up" more than once while parsing a single packet, with the same contents in each?  Or to be used in two different parser states, even if not more than once per packet?

If so, the current definition of value_set in the spec makes it straightforward to do so, even if the `select` expressions contain _different_ expressions.  P4_16 table definition syntax would make that somewhat awkward to do.

**Comment by antoninbas** [comment]
That's true. But a table-like syntax also makes it easier to generate control-plane names.
When the type parameter of the `value_set` is a struct, it's not so bad, but what about this case: `value_set<tuple<bit<8>, bit<16> > <`
 * do I consider that I have one exact match on `tuple<bit<8>, bit<16> >` or two exact matches on `bit<8>` and `bit<16>` respectively?
 * in either case, how do I name the match(es)?

**Comment by antoninbas** [comment]
The spec states that:

> If the value_set takes a struct as type parameter, the runtime API can use the struct field names to name the objects in the value set.

So maybe for P4Runtime, we can add the restriction that the type parameter *must* be a struct. Then we solve the symmetry issue + the naming issue.

**Comment by jafingerhut** [comment]
At least for P4Runtime v1.0, specifying how this works for a struct type, and saying that it is not supported to use anything else in a value_set, seems reasonable to me.  If we come up with a way to support other use cases of interest later, they can be described in a later version of the P4Runtime spec.
